waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,716 @@
+    -- of this, avoid making alternatives between rules that can be ambiguous
+    -- in this way.
+    --
+    -- TODO: Try running each rule and use backtracking on failure
+    Alternative GrammarRule GrammarRule
+  deriving (Eq, Ord, Show)
+
+-- | @rule1 '<>' rule2@ is a parser that runs first @rule1@, then @rule2@.
+instance Semigroup GrammarRule where
+  -- "Succeed" is identity of <>. The second pattern does affect the result tree,
+  -- but only by changing where trivia tokens are placed in the tree, which is
+  -- not an issue.
+  Succeed <> rule = rule
+  rule <> Succeed = rule
+  first <> second = Chain first second
+
+-- | @mempty@ is a parser that matches an empty string, or a string of only
+-- trivia tokens.
+instance Monoid GrammarRule where
+  mempty = Succeed
+
+-- | Run a "GrammarRule" on an input of tokens. Returns a tuple @(errors, nodes)@.
+parse :: [Token] -> GrammarRule -> ([ParseError], [SyntaxNode])
+parse tokens grammar =
+  let initialState =
+        ParseState
+          { pstateInput = tokens,
+            pstateCurrentLevelNodes = [],
+            pstateErrors = [],
+            pstateNextTokenOffset = 0,
+            pstatePrevTokenOffset = 0,
+            pstateFollowingTokens = []
+          }
+   in case runState (runExceptT (parseRule grammar)) initialState of
+        (Left err, state) -> case err of
+          Unwind _ -> error "Unwind at top-level GrammarRule (impossible, this is a bug in Wasp.Backend.ConcreteParser.Internal)"
+          ParseError perr ->
+            let errs = pstateErrors state ++ [perr]
+                newNode = errorNode perr False
+                nodes = pstateCurrentLevelNodes state ++ [newNode]
+             in (errs, nodes)
+        (_, state) -> (pstateErrors state, pstateCurrentLevelNodes state)
+
+data ParseState = ParseState
+  { -- | Remaining tokens in input.
+    pstateInput :: [Token],
+    -- | "SyntaxNode"s processed so far at the current level of the CST. An
+    -- initial top-level is created when the parser is started and each "Group"
+    -- rule creates a new level.
+    pstateCurrentLevelNodes :: [SyntaxNode],
+    -- | Errors collected that will be reported.
+    pstateErrors :: [ParseError],
+    -- | 0-based offset from the beginning of the source file, in characters,
+    -- of the start of the next token (@head . `pstateInput`@).
+    pstateNextTokenOffset :: !Int,
+    -- | 0-based offset from the beginning of the source file, in characters,
+    -- of the start of the last consumed token
+    pstatePrevTokenOffset :: !Int,
+    -- | Stack of possible first tokens, collected when we enter a group as the
+    -- first grammar rule in a "Chain".
+    --
+    -- A member @None@ represents EOF.
+    pstateFollowingTokens :: [TokenSet]
+  }
+  deriving (Eq, Show, Ord)
+
+data ParseException
+  = -- | @Unwind tokenKind@ is used as a signal during error recovery. It is used
+    -- to implement the unwinding behavior described at the top of the module.
```

Comment:
I guess I was trying to do exactly that, detach it from the strategy itself.
Because when you see Unwind, you can't understand what it is without understanding the rest of the strategy. But if you see RecoverableError, you already know something about it, even if you don't know how recovery works.
I wouldn't say it implies `ParseError` is not recoverable, it just means we don't know if it is recoverable or not. But well, if you thought it implies, then so will some others hm! That could be explained in comments.

Ok, let's stick with Unwind if you feel better about it, since you already have a lot of comments and code using "unwind" term!